provider/src/cache.ts

```diff
@@ -151,3 +55,151 @@ export const storeContractStatusInKV = async (
     console.error('Error storing contract status in KV:', error)
   }
 }
+
+/**
+ * Get a cached response from KV storage
+ * @param request The RPC request
+ * @param chainId The chain ID
+ * @param env The environment variables
+ * @returns The cached response or null if not found
+ *
+ * First attempts to retrieve and parse as JSON directly. If that fails,
+ * falls back to retrieving as text and parsing manually. If the text
+ * parsing also fails, deletes the corrupted cache entry.
+ */
+export const getResponseFromKV = async (
+  request: RpcRequest,
+  chainId: string,
+  env: Env
+): Promise<any> => {
+  if (!env.REQUEST_CACHE) {
+    return null
+  }
+
+  const cacheKey = generateRequestCacheKey(request)
+
+  try {
+    // Try to get and parse as JSON
+    const cached = await env.REQUEST_CACHE.get(cacheKey, 'json')
+    if (cached) {
+      return cached
+    }
+  } catch (error) {
+    // If JSON parsing fails, try simple text retrieval and parse it
+    try {
+      const rawText = await env.REQUEST_CACHE.get(cacheKey, 'text')
+      if (rawText) {
+        try {
+          return JSON.parse(rawText)
+        } catch {
+          // If still can't parse, delete the corrupted entry
+          await env.REQUEST_CACHE.delete(cacheKey).catch((err) => {
+            console.error(`Error deleting corrupted cache entry: ${err}`)
+          })
+        }
+      }
+    } catch (error) {
+      console.error(
+        `Error reading from cache: ${error instanceof Error ? error.message : String(error)}`
+      )
+    }
+  }
+
+  return null
+}
+
+/**
+ * Cache a response
+ * @param request The RPC request
+ * @param response The response to cache
+ * @param chainId The chain ID
+ * @param env The environment variables
+ */
+export const storeResponseInKV = async (
+  request: RpcRequest,
+  response: any,
+  chainId: string,
+  env: Env
+): Promise<void> => {
+  if (!env.REQUEST_CACHE || !response?.result) {
+    return
+  }
+  const cacheKey = generateRequestCacheKey(request)
+  const ttl = getCacheTTL(env)
+
+  try {
+    const cacheValue = { result: response.result }
+    await env.REQUEST_CACHE.put(cacheKey, JSON.stringify(cacheValue), {
+      expirationTtl: ttl,
+    })
+  } catch (error) {
+    console.error(
+      `Error writing to cache: ${error instanceof Error ? error.message : String(error)}`
+    )
+  }
+}
+
+/**
+ * Determine if a request should be cached
+ * @param request The RPC request
+ * @param chainId The chain ID
+ * @returns True if the request should be cached, false otherwise
+ */
+export const shouldStore = (request: RpcRequest, chainId: string): boolean => {
+  if (!request.method) return false
+
+  // check if it's a name resolution request
+  if (
+    request.method.toLowerCase() === 'eth_call' &&
+    isNameResolutionRequest(request, chainId)
```

Comment:
`request.method.toLowerCase() === 'eth_call'` shou;d be _inside_ `isNameResolutionRequest` for simplicity